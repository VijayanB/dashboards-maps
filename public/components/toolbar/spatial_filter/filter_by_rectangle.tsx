/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { EuiButtonIcon, EuiContextMenu, EuiPanel, EuiPopover } from '@elastic/eui';
import { FilterInputPanel } from './filter_input_panel';
import {
  DRAW_FILTER_CANCEL,
  DRAW_FILTER_RECTANGLE,
  DRAW_FILTER_RECTANGLE_DEFAULT_LABEL,
  DRAW_FILTER_SHAPE_TITLE,
  DRAW_FILTER_SPATIAL_RELATIONS,
  DrawFilterProperties,
  FILTER_DRAW_MODE,
} from '../../../../common';

interface FilterByRectangleProps {
  setDrawFilterProperties: (properties: DrawFilterProperties) => void;
  mode: FILTER_DRAW_MODE;
}

const isFilterActive = (mode: FILTER_DRAW_MODE): boolean => {
  return mode === FILTER_DRAW_MODE.RECTANGLE;
};

export const FilterByRectangle = ({ setDrawFilterProperties, mode }: FilterByRectangleProps) => {
  const [isPopoverOpen, setPopover] = useState(false);

  const onClick = () => {
    if (isFilterActive(mode)) {
      setDrawFilterProperties({
        mode: FILTER_DRAW_MODE.NONE,
      });
      return;
    }
    setPopover(!isPopoverOpen);
  };

  const closePopover = () => {
    setPopover(false);
  };

  const onSubmit = (input: { relation: string; label: string; mode: FILTER_DRAW_MODE }) => {
    setDrawFilterProperties({
      mode: input.mode,
      relation: input.relation,
      filterLabel: input.label,
    });
    closePopover();
  };

  const panels = [
    {
      id: 0,
      title: DRAW_FILTER_SHAPE_TITLE,
      content: (
        <FilterInputPanel
          drawLabel={DRAW_FILTER_RECTANGLE}
          defaultFilterLabel={DRAW_FILTER_RECTANGLE_DEFAULT_LABEL}
          relations={DRAW_FILTER_SPATIAL_RELATIONS}
          onSubmit={onSubmit}
          mode={FILTER_DRAW_MODE.RECTANGLE}
        />
      ),
    },
  ];

  const drawRectangleButton = (
    <EuiPanel paddingSize="none" className="spatialFilterToolbar__shape">
      <EuiButtonIcon
        isSelected={isFilterActive(mode)}
        display={isFilterActive(mode) ? 'fill' : 'empty'}
        aria-pressed={isFilterActive(mode)}
        color={'primary'}
        size={'s'}
        iconType={'vector'}
        onClick={onClick}
        aria-label={'draw_filter_rectangle'}
        title={isFilterActive(mode) ? DRAW_FILTER_CANCEL : DRAW_FILTER_RECTANGLE}
      />
    </EuiPanel>
  );
  return (
    <EuiPopover
      id="drawRectangleId"
      button={drawRectangleButton}
      isOpen={isPopoverOpen}
      closePopover={closePopover}
      panelPaddingSize="none"
      anchorPosition="leftUp"
      data-test-subj="drawRectanglePopOver"
    >
      <EuiContextMenu initialPanelId={0} panels={panels} />
    </EuiPopover>
  );
};
