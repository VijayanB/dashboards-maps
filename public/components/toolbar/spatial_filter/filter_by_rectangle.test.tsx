/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { shallow } from 'enzyme';
import React from 'react';
import { FilterByRectangle } from './filter_by_rectangle';

it('renders filter by rectangle button', () => {
  const mockCallback = jest.fn();
  const component = shallow(
    <FilterByRectangle setDrawFilterProperties={mockCallback} isDrawActive={false} />
  );
  expect(component).toMatchSnapshot();
});

it('renders filter by rectangle in middle of drawing', () => {
  const mockCallback = jest.fn();
  const component = shallow(
    <FilterByRectangle setDrawFilterProperties={mockCallback} isDrawActive={true} />
  );
  expect(component).toMatchSnapshot();
});
