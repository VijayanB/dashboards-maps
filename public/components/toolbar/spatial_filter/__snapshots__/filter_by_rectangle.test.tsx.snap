// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders filter by rectangle button 1`] = `
<EuiPopover
  anchorPosition="leftUp"
  button={
    <EuiPanel
      className="spatialFilterToolbar__shape"
      paddingSize="none"
    >
      <EuiButtonIcon
        aria-label="draw_filter_rectangle"
        aria-pressed={false}
        color="primary"
        display="empty"
        iconType="vector"
        isSelected={false}
        onClick={[Function]}
        size="s"
        title="Draw Rectangle"
      />
    </EuiPanel>
  }
  closePopover={[Function]}
  data-test-subj="drawRectanglePopOver"
  display="inlineBlock"
  hasArrow={true}
  id="drawRectangleId"
  isOpen={false}
  ownFocus={true}
  panelPaddingSize="none"
>
  <EuiContextMenu
    initialPanelId={0}
    panels={
      Array [
        Object {
          "content": <FilterInputPanel
            defaultFilterLabel="rectangle"
            drawLabel="Draw Rectangle"
            mode="rectangle"
            onSubmit={[Function]}
            relations={
              Array [
                "intersects",
                "disjoint",
                "within",
              ]
            }
          />,
          "id": 0,
          "title": "DRAW SHAPE",
        },
      ]
    }
    size="m"
  />
</EuiPopover>
`;

exports[`renders filter by rectangle in middle of drawing 1`] = `
<EuiPopover
  anchorPosition="leftUp"
  button={
    <EuiPanel
      className="spatialFilterToolbar__shape"
      paddingSize="none"
    >
      <EuiButtonIcon
        aria-label="draw_filter_rectangle"
        aria-pressed={false}
        color="primary"
        display="empty"
        iconType="vector"
        isSelected={false}
        onClick={[Function]}
        size="s"
        title="Draw Rectangle"
      />
    </EuiPanel>
  }
  closePopover={[Function]}
  data-test-subj="drawRectanglePopOver"
  display="inlineBlock"
  hasArrow={true}
  id="drawRectangleId"
  isOpen={false}
  ownFocus={true}
  panelPaddingSize="none"
>
  <EuiContextMenu
    initialPanelId={0}
    panels={
      Array [
        Object {
          "content": <FilterInputPanel
            defaultFilterLabel="rectangle"
            drawLabel="Draw Rectangle"
            mode="rectangle"
            onSubmit={[Function]}
            relations={
              Array [
                "intersects",
                "disjoint",
                "within",
              ]
            }
          />,
          "id": 0,
          "title": "DRAW SHAPE",
        },
      ]
    }
    size="m"
  />
</EuiPopover>
`;
